streams/src/test/java/org/apache/kafka/streams/state/internals/RocksDBVersionedStoreTest.java

```diff
@@ -0,0 +1,435 @@
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL + 30, "vp20", SEGMENT_INTERVAL + 20);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL + 12);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 1);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 2, "vn2", SEGMENT_INTERVAL - 2);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 5);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 6, "vn6", SEGMENT_INTERVAL - 6);
+    }
+
+    @Test
+    public void shouldPutNonLatestTombstoneIntoNewSegmentWithValidTo() {
+        putToStore("k", "vp30", SEGMENT_INTERVAL + 30);
+        putToStore("k", null, SEGMENT_INTERVAL - 10); // this put should result in tombstone with validTo=SEGMENT_INTERVAL+30
+        putToStore("k", "vn5", SEGMENT_INTERVAL - 5);
+        putToStore("k", "vn1", SEGMENT_INTERVAL - 1);
+        putToStore("k", null, SEGMENT_INTERVAL - 2);
+
+        verifyGetValueFromStore("k", "vp30", SEGMENT_INTERVAL + 30);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL + 10, "vn1", SEGMENT_INTERVAL - 1);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 1, "vn1", SEGMENT_INTERVAL - 1);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 2);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 5, "vn5", SEGMENT_INTERVAL - 5);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 10);
+    }
+
+    @Test
+    public void shouldGetFromOlderSegments() {
+        // use a different key to create three different segments
+        putToStore("ko", null, SEGMENT_INTERVAL - 10);
+        putToStore("ko", null, 2 * SEGMENT_INTERVAL - 10);
+        putToStore("ko", null, 3 * SEGMENT_INTERVAL - 10);
+
+        // return null after visiting all segments
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 20);
+
+        // insert data to create non-empty (first) segment
+        putToStore("k", "v", SEGMENT_INTERVAL - 20);
+        putToStore("k", null, SEGMENT_INTERVAL - 10);
+
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 30);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 15, "v", SEGMENT_INTERVAL - 20);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 5);
+
+        // insert data to create non-empty (third) segment
+        putToStore("k", "v2", 3 * SEGMENT_INTERVAL - 20);
+        putToStore("k", null, 3 * SEGMENT_INTERVAL - 10);
+
+        // presence of non-empty later segment does not affect results of getting from earlier segment
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 30);
+        verifyTimestampedGetValueFromStore("k", SEGMENT_INTERVAL - 15, "v", SEGMENT_INTERVAL - 20);
+        verifyTimestampedGetNullFromStore("k", SEGMENT_INTERVAL - 5);
+    }
+
+    @Test
+    public void shouldNotGetExpired() {
```

Comment:
Should we also test the `validTo`-update-case via an "out of order" put (regular record or null) leading to an existing record begin expired?